google/cloud/datastore/query.py

```diff
@@ -804,46 +813,69 @@ def _next_page(self):
             "project_id": self._query.project,
             "partition_id": partition_id,
             "read_options": read_options,
-            "query": query_pb,
+            "query": self._build_protobuf(),
         }
+        if self._query._explain_options:
+            request["explain_options"] = self._query._explain_options._to_dict()
 
         helpers.set_database_id_to_request(request, self.client.database)
 
-        response_pb = self.client._datastore_api.run_query(
-            request=request,
-            **kwargs,
-        )
+        response_pb = None
 
-        while (
+        while response_pb is None or (
             response_pb.batch.more_results == _NOT_FINISHED
-            and response_pb.batch.skipped_results < query_pb.offset
+            and response_pb.batch.skipped_results < request["query"].offset
         ):
-            # We haven't finished processing. A likely reason is we haven't
-            # skipped all of the results yet. Don't return any results.
-            # Instead, rerun query, adjusting offsets. Datastore doesn't process
-            # more than 1000 skipped results in a query.
-            old_query_pb = query_pb
-            query_pb = query_pb2.Query()
-            query_pb._pb.CopyFrom(old_query_pb._pb)  # copy for testability
-            query_pb.start_cursor = response_pb.batch.skipped_cursor
-            query_pb.offset -= response_pb.batch.skipped_results
-
-            request = {
-                "project_id": self._query.project,
-                "partition_id": partition_id,
-                "read_options": read_options,
-                "query": query_pb,
-            }
-            helpers.set_database_id_to_request(request, self.client.database)
+            if response_pb is not None:
+                # We haven't finished processing. A likely reason is we haven't
+                # skipped all of the results yet. Don't return any results.
+                # Instead, rerun query, adjusting offsets. Datastore doesn't process
+                # more than 1000 skipped results in a query.
+                new_query_pb = query_pb2.Query()
+                new_query_pb._pb.CopyFrom(request["query"]._pb)  # copy for testability
+                new_query_pb.start_cursor = response_pb.batch.skipped_cursor
+                new_query_pb.offset -= response_pb.batch.skipped_results
+                request["query"] = new_query_pb
 
             response_pb = self.client._datastore_api.run_query(
-                request=request,
-                **kwargs,
+                request=request.copy(), **kwargs
```

Comment:
Is `request.copy()` here also for testability? `copy()` is only a shallow copy for dict, do we need to make a deep copy here?

Reply:
Yeah, IIRC it's just for testability. 

We want to be able to make assertions about the arguments passed in to each call, which doesn't work if the same request instance is used for each request